rust/loqui_client/src/connection_handler.rs

```
@@ -0,0 +1,306 @@
+                let request = Request {
+                    payload,
+                    sequence_id,
+                    flags: make_flags(compressed),
+                };
+                Some(request.into())
+            }
+            Err(e) => {
+                error!("Failed to encode payload. error={:?}", e);
+                None
+            }
+        }
+    }
+
+    fn handle_response(&mut self, response: Response, transport_options: &TransportOptions) {
+        let Response {
+            flags,
+            sequence_id,
+            payload,
+        } = response;
+        match self.waiters.remove(&sequence_id) {
+            Some(waiter_tx) => {
+                let response = self.config.encoder.decode(
+                    transport_options.encoding,
+                    is_compressed(flags),
+                    payload,
+                );
+                if let Err(_e) = waiter_tx.send(response) {
+                    error!("Waiter is no longer listening.")
+                }
+            }
+            None => {
+                error!("No waiter for sequence_id. sequence_id={:?}", sequence_id);
+            }
+        }
+    }
+
+    fn handle_error(&mut self, error: ErrorFrame) {
+        let ErrorFrame {
+            sequence_id,
+            payload,
+            ..
+        } = error;
+        match self.waiters.remove(&sequence_id) {
+            Some(waiter_tx) => {
+                // payload is always a string
+                match String::from_utf8(payload) {
+                    Ok(reason) => {
+                        if let Err(_e) = waiter_tx.send(Err(LoquiError::InternalServerError {
+                            error: err_msg(reason),
+                        }
+                        .into()))
+                        {
+                            error!("Waiter is no longer listening.")
+                        }
+                    }
+                    Err(e) => {
+                        error!(
```

We should return this error to the waiter_Tx too.